File: scratch/excel_generator.py

import pandas as pd
import psycopg2
import os

db_url = os.getenv("DATABASE_URL", "postgresql://aba_admin:Pr0s1s.2026@localhost:5432/aba_supervision")

def get_tables():
    conn = psycopg2.connect(db_url)
    cur = conn.cursor()
    cur.execute("SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname = 'public' ORDER BY tablename;")
    tables = [t[0] for t in cur.fetchall()]
    cur.close()
    conn.close()
    return tables

def get_metadata(table_name):
    conn = psycopg2.connect(db_url)
    cur = conn.cursor()
    query = f"""
        SELECT 
            c.column_name, 
            c.data_type, 
            c.is_nullable,
            (SELECT 'PK' FROM information_schema.key_column_usage k 
             WHERE k.table_name = c.table_name AND k.column_name = c.column_name AND k.table_schema = c.table_schema
             AND EXISTS (SELECT 1 FROM information_schema.table_constraints tc 
                         WHERE tc.constraint_name = k.constraint_name AND tc.constraint_type = 'PRIMARY KEY')) as is_pk
        FROM information_schema.columns c
        WHERE c.table_name = '{table_name}' AND c.table_schema = 'public'
        ORDER BY c.ordinal_position;
    """
    cur.execute(query)
    meta = cur.fetchall()
    cur.close()
    conn.close()
    return meta

def get_real_data(table_name, limit=5):
    conn = psycopg2.connect(db_url)
    try:
        df = pd.read_sql_query(f'SELECT * FROM "{table_name}" LIMIT {limit}', conn)
        return df
    except:
        return pd.DataFrame()
    finally:
        conn.close()

def main():
    tables = get_tables()
    output_file = "CARGA_MASIVA_MAESTRA_ABA.xlsx"
    
    with pd.ExcelWriter(output_file, engine='openpyxl') as writer:
        for table in tables:
            print(f"Processing {table}...")
            meta = get_metadata(table)
            data = get_real_data(table)
            
            # Create a guide row from metadata
            guide = []
            headers = []
            for m in meta:
                headers.append(m[0])
                desc = m[1].upper()
                if m[3] == 'PK': desc += " (ID)"
                if m[2] == 'NO': desc += " (REQ)"
                guide.append(desc)
            
            # Create a DataFrame for the sheet
            # Row 0: Headers
            # Row 1: Guide
            # Row 2+: Data
            
            df_guide = pd.DataFrame([guide], columns=headers)
            if not data.empty:
                # Align data columns with headers just in case
                data = data.reindex(columns=headers)
                df_final = pd.concat([df_guide, data], ignore_index=True)
            else:
                df_final = df_guide
                
            # Shorten table name for sheet tab (max 31 chars)
            sheet_name = table[:31]
            df_final.to_excel(writer, sheet_name=sheet_name, index=False)
            
    print(f"SUCCESS: {output_file} generated.")

if __name__ == "__main__":
    main()

File: scratch/excel_generator.js

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const dbUrl = process.env.DATABASE_URL || "postgresql://aba_admin:Pr0s1s.2026@localhost:5432/aba_supervision";

function runPsql(query) {
    try {
        const escapedQuery = query.replace(/"/g, '\\"');
        const cmd = `psql "${dbUrl}" -t -A -F "|" -c "${escapedQuery}"`;
        const result = execSync(cmd, { encoding: 'utf8' }).trim();
        return result ? result.split('\n').map(line => line.split('|')) : [];
    } catch (err) {
        return [];
    }
}

function getTables() {
    return runPsql("SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname = 'public' ORDER BY tablename;").map(t => t[0]);
}

function getTableMetadata(tableName) {
    const query = `
        SELECT 
            c.column_name, 
            c.data_type, 
            c.is_nullable,
            (SELECT 'PK' FROM information_schema.key_column_usage k 
             WHERE k.table_name = c.table_name AND k.column_name = c.column_name AND k.table_schema = c.table_schema
             AND EXISTS (SELECT 1 FROM information_schema.table_constraints tc 
                         WHERE tc.constraint_name = k.constraint_name AND tc.constraint_type = 'PRIMARY KEY')) as is_pk
        FROM information_schema.columns c
        WHERE c.table_name = '${tableName}' AND c.table_schema = 'public'
        ORDER BY c.ordinal_position;
    `;
    return runPsql(query);
}

function getRealData(tableName, limit = 5) {
    return runPsql(`SELECT * FROM "${tableName}" LIMIT ${limit};`);
}

async function main() {
    const tables = getTables();
    console.log(`Generating Master Excel (XML format) with ${tables.length} sheets...`);
    
    let xml = `<?xml version="1.0"?>
<?mso-application progid="Excel.Sheet"?>
<Workbook xmlns="urn:schemas-microsoft-com:office:spreadsheet"
 xmlns:o="urn:schemas-microsoft-com:office:office"
 xmlns:x="urn:schemas-microsoft-com:office:excel"
 xmlns:ss="urn:schemas-microsoft-com:office:spreadsheet"
 xmlns:html="http://www.w3.org/TR/REC-html40">
 <Styles>
  <Style ss:ID="header">
   <Font ss:Bold="1" ss:Color="#FFFFFF"/>
   <Interior ss:Color="#4F81BD" ss:Pattern="Solid"/>
  </Style>
  <Style ss:ID="guide">
   <Font ss:Italic="1" ss:Color="#555555"/>
   <Interior ss:Color="#DCE6F1" ss:Pattern="Solid"/>
  </Style>
 </Styles>`;

    for (const table of tables) {
        console.log(`  Adding sheet: ${table}...`);
        const meta = getTableMetadata(table);
        const data = getRealData(table);
        
        xml += `\n <Worksheet ss:Name="${table.substring(0, 31)}">\n  <Table>`;
        
        // Headers row
        xml += `\n   <Row ss:StyleID="header">`;
        meta.forEach(m => {
            xml += `\n    <Cell><Data ss:Type="String">${m[0]}</Data></Cell>`;
        });
        xml += `\n   </Row>`;
        
        // Guide row
        xml += `\n   <Row ss:StyleID="guide">`;
        meta.forEach(m => {
            let guide = m[1].toUpperCase();
            if (m[3] === 'PK') guide += " (ID)";
            if (m[2] === 'NO') guide += " (REQ)";
            xml += `\n    <Cell><Data ss:Type="String">${guide}</Data></Cell>`;
        });
        xml += `\n   </Row>`;
        
        // Data rows
        data.forEach(row => {
            xml += `\n   <Row>`;
            row.forEach(val => {
                const escaped = String(val || '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
                xml += `\n    <Cell><Data ss:Type="String">${escaped}</Data></Cell>`;
            });
            xml += `\n   </Row>`;
        });
        
        xml += `\n  </Table>\n </Worksheet>`;
    }
    
    xml += `\n</Workbook>`;
    
    fs.writeFileSync('CARGA_MASIVA_MAESTRA_ABA.xls', xml);
    console.log(`\nSUCCESS: CARGA_MASIVA_MAESTRA_ABA.xls generated.`);
}

main();
